### DISCUSSION TITLE
Color token naming convention

### DISCUSSION THREAD

**Discussion by aagonzales** [carbon-design-system/carbon#7827] [Ideas]
# Goal
As part of our next major release we want to update the names of the color tokens to users better understand how to apply color without needing to frequently reference a guide. We want the names to clearly articulate the purpose of the color and where/how to apply it.

**Note: These token names and lists are based on the assumption that we will be implementing the [inline theming proposal](https://github.com/carbon-design-system/carbon/discussions/7743#discussioncomment-350979).** 

### Existing convention 
Currently, our [existing token names](https://carbondesignsystem.com/guidelines/color/usage) are too abstract and only use numbers to distinguish between roles. For example, with `text-01` the user knows its a text color but what kind of text? 

There is also no real convention to the names and names have become inconsistent. 

# Two proposed naming conventions
**Conventions**
- A. [style modifier*]-[element]-[role]-[state]: [See comment](https://github.com/carbon-design-system/carbon/discussions/7827#discussioncomment-379267)
- B. [property]-[role]-[style modifier]-[state]: [See comment](https://github.com/carbon-design-system/carbon/discussions/7827#discussioncomment-379278)

**Terms**:
-  **Element**: the part of the ui the token is being applied to, (ie border, button). 
- **Property**: as in a css property
- **Role**: the different specific visual or functional purpose of a single element.
    - Roles can enforce hierarchy (ie primary, secondary) or refine the usage of the element (ie subtle, danger).
- **Style modifier**: `*` the style modifier is not a required in every token. It is only present in a few token categories. It is used to help group tokens that are used as a set (inverse) or interchangeably (support).
- **State**: the interactive state (ie hover, active) being applied to its token parent.

*Overall we are (mostly) removing the terminating number labeling `-01` `-02` etc*

# Request for feedback:
- Which token naming convention do you prefer, [A](https://github.com/carbon-design-system/carbon/discussions/7827#discussioncomment-379267) or [B](https://github.com/carbon-design-system/carbon/discussions/7827#discussioncomment-379278)? Comment in the threads or 👍.
- What should we do about disabled token names? https://github.com/carbon-design-system/carbon/discussions/7827#discussioncomment-379330 
- Should we combine the old field-01 and ui-01 for this version or keep them separate? https://github.com/carbon-design-system/carbon/discussions/7827#discussioncomment-379351
- Should we add a `color-` prefix? https://github.com/carbon-design-system/carbon/discussions/7827#discussioncomment-379372
- What should the new ui-01 name be? https://github.com/carbon-design-system/carbon/discussions/7827#discussioncomment-379397
- Do any of the names not make sense to you or would you not understand how to apply it? If so please leave a comment about which one and any name change suggestions.

# Conclusion
https://github.com/carbon-design-system/carbon/discussions/7827#discussioncomment-558179

**Comment by aagonzales** [comment]
## A. [style modifier*]-[element]-[role]-[state]
[View full token list](https://www.notion.so/carbondesignsystem/Proposals-2a-List-of-to-be-tokens-e997ba016e3a48ddb1fb96346eb1e929)

- This convention keeps the naming structure similar to existing v10 names for there will be a familiarity in migration.
- It has shorter and more direct names. 
- Style modifier creates a category among tokens that are used together or interchangeably. An example of a style modifier is "inverse" or "support"
- Tokens can have one word element names when calling out the primary or only use of color applied to an element. (ie background, highlight, layer, field)

**Examples:**
- `[element]`
   - `$background`
   - `$layer`
- `[element]-[role]`
   - `$border-subtle`
   - `$text-helper`
- `[style modifier]-[element]`
   - `$inverse-background`   
   - `$interactive-background`
-  `[style modifier]-[element]-[role]`
   - `$inverse-support-success`
- `[element]-[state]`
   - `$background-hover`
   - `$layer-active`
- `[element]-[role]-[state]`
   - `$button-primary-active`
   - `$border-subtle-selected`

![image](https://user-images.githubusercontent.com/11670886/108270304-46d15600-713d-11eb-91a0-f80a223a6be3.png)

**Reply by tessarodes** [reply]
I guarantee I'm missing a major implication but centering around an "element" feels natural to me.

**Reply by tw15egan** [reply]
I like having the modifier come first, easy to group and recognize similar tokens i.e inverse

**Comment by aagonzales** [comment]
## B. [property]-[role]-[style modifier]-[state]
[View full token list](https://www.notion.so/carbondesignsystem/Proposals-2b-List-of-to-be-tokens-a6bb4311a5724404acb037f5e9d2f405)

- This naming convention is focused more on the (css?) property the value is applied too. It introduces new terms like "background" and "fill"
- However, some tokens will still begin with the "element" name versus a "property" to better aid in usage understanding.
- This results in longer compound names but clearer usage. 
- Terms like "support" as removed in favor of using a property-role combination instead. For example `support-01` is now `fill-error`
- The property or element always comes first
   - The exception is component level tokens in which case the component name would come first, ie `toggle-off`

**Examples:**
- `[property]-[role]`
   - `$background-theme`
   - `$border-subtle`
   - `$text-helper`
   - `$fill-warning`
- `[property]-[style modifier]`
   - `$background-inverse`   
-  `[property]-[role]-[style modifier]`
   - `$fill-success-inverse`
   - `$button-primary-danger`
- `[property]-[role]-[state]`
   - `$background-theme-hover`
   - `$background-component-active`

![image](https://user-images.githubusercontent.com/11670886/108270235-2dc8a500-713d-11eb-89fd-46a67958c32e.png)

**Reply by joshblack** [reply]
This proposal is super interesting, the table really shows how useful the names are for understanding how/where to apply a token.

I think the only one that stood out was the `fill` one, specifically since SVGs could be styled using `stroke` or `fill` (although our whole set now uses `fill`). Are these `fill-*` tokens specifically for SVGs, or could they be applied to other elements on a page?

**Reply by aagonzales** [reply]
Fill can be applied to more than just svg. If it was an svg only color then it would be an `icon-` token. A lot of these "fill" colors have a clear role but I wasn't sure what property or element to assign them. Like I took away "support" in favor of "fill" but we can bring back some of those terms if we think it works better.

**Reply by laurenmrice** [reply]
I am a little torn between the two options but I think I am liking B more. I like the simplicity of A. B has a more complex naming convention but it explains the usage of each token better, which is what we were originally trying to achieve.

**Reply by laurenmrice** [reply]
The `fill-` naming conventions of the B set are ambiguous to me (similar reasons as to what Josh said above^). I feel like we could add `support-` back in for the support related tokens. For `fill-interactive` (interactive-04) it could just be `interactive` or something else?

**Reply by aagonzales** [reply]
`support-` is just not a property so breaks with the convention. But maybe it's better to have the familiar more easily understood term versus forcing a convention that is a worse experience.

**Reply by yumeforever** [reply]
I don't have a particular preference here, but having the element/root coming first seems more natural to me.
I like to be closer to the code implementation style too.

**Comment by aagonzales** [comment]
## Question: disabled tokens
Should we break apart the abstract disabled tokens into more specific names? This would result in more disabled tokens going from 3 to about ~8. Or should we just keep it simple with abstract tokens.

![image](https://user-images.githubusercontent.com/11670886/108272919-e17f6400-7140-11eb-8a66-761bfcf71e61.png)

**Reply by laurenmrice** [reply]
I would be fine with having different specific disabled tokens if it makes it easier for people to use them.

*The optional naming of `text-white` and `icon-white` for `-on-color` that you have in the notion tables seem odd if we end up introducing disabled tokens for them (it is the only token right now that would be specifying an actual color in the token name). Probably best to keep those as `on-color` like you are showing in this image here.

**Reply by tessarodes** [reply]
Interesting! Sure. I don't think it's confusing to break it out and could add clarity.

**Reply by aagonzales** [reply]
I was actually on the side of keeping the abstract because its less work lol but breaking them apart is probably the better systematic way to do it.

**Reply by yumeforever** [reply]
+1 it just make sense if you follow any of the other rationale, helps making things more descriptive

**Comment by aagonzales** [comment]
## Question: field token
Should we combine the traditional `field` token with the main "layer"/"background-component" token? In our color modal the fields (old field-01) are the same color as the primary layer color (old ui-01). 

Combining would result in less tokens but we would lose the ability to change field color separate from other ui containers. This is mostly a future concern or a "white label theme" concern. 

Note: we no longer need two field tokens with inline theming.

![image](https://user-images.githubusercontent.com/11670886/108274543-0a085d80-7143-11eb-8725-4330d6afd61f.png)

**Reply by joshblack** [reply]
It might be nice to keep the field tokens for the use-cases you described, the alternative if we needed to distinguish the colors seems like it would be component tokens per form input component and that seems like a drag compared to just specifying field tokens. Let me know if there is an easier way though, I could easily be misunderstanding 👀

**Reply by laurenmrice** [reply]
I think we could just keep the `field` (`layer`/`background-component`) token for now and see how it plays out for awhile. If we eventually feel it is still redundant we could get rid of it. That is a good point about keeping it for future possibilities of “white labeling”.

**Reply by aagonzales** [reply]
If we keep both (which I do like), I think we'll just need to be more consistent about when you use `field`. I think in some places we use it in the field but not in other component areas, like here in dropdown. 

![image](https://user-images.githubusercontent.com/11670886/109336432-60575980-7831-11eb-8d07-bb18f4ccf149.png)

Theoretically, should the list box background also be using field-01 then?

**Reply by yumeforever** [reply]
Keeping both and using the layer 01/02/03 makes sense to me

**Comment by aagonzales** [comment]
## Question: color prefix
Should we prefix the tokens with `color-`? So tokens would read like `$color-button-primary` or `$color-background`. 

If I remember correctly, @joshblack said we could make this essentially part of the file structure.

**Reply by joshblack** [reply]
@aagonzales yeah, we could definitely have this be related to the hierarchy/category/file versus having to be in the name.

**Reply by laurenmrice** [reply]
I don’t have a strong opinion about this but things I have noted: 
- Adding a prefix will make the token names longer and add more repetition, which could make them harder to read or choose from.
- The token names without the prefix already resonate as a `color` token to me. I would be interested to know what other developers think about this.

**Reply by tw15egan** [reply]
> Adding a prefix will make the token names longer and add more repetition, which could make them harder to read or choose from.

That's my concern as well. I wonder if we could alias them or something so that they work with / without a `color` prefix, if that is desired by some.

> The token names without the prefix already resonate as a color token to me.

I agree. Especially since a dev would most likely be using in a context like so

```scss
body {
   background-color: $layer
}
```

**Reply by yumeforever** [reply]
I also dont see the need for the extra color-

**Comment by aagonzales** [comment]
## Question: new ui-01 name
What should we call the new ui-01 token? In proposal A we call it `layer` and proposal `B` we call it `background-component`. Are there other names out there that may fit better? Material and Polaris call it a "surface". 

![image](https://user-images.githubusercontent.com/11670886/108275914-05dd3f80-7145-11eb-84bd-b5abcdfd5a05.png)

**Reply by joshblack** [reply]
Layer is a fun name and it seems like it could be simple to understand since it can be mapped into design tooling with layers. It also would play well with the inline theming / layering proposal where-as a term like surface seems like it might not be as simple to guess what its function is.

**Reply by laurenmrice** [reply]
I like having the name `layer` for A. For B, another name we could try is `background-layer`?

**Reply by aagonzales** [reply]
Yeah I struggled with the name for B. My thinking was it always a "layer" when used for a component? I guess so? I think `background-layer` would work fine too. I was really torn with this one so hopefully other chime in with their thoughts.

**Reply by tw15egan** [reply]
I like `layer` as well. Another variation could be `base-layer`?

**Reply by aagonzales** [reply]
Oh `base` is an interesting term I hadn't thought about before.

**Reply by yumeforever** [reply]
With base-layer or background-layer, would it add the number too?
I.e.: `base-layer-01` `base-layer-02`?
I prefer the Options B so I'd be ok with the shortest one, like base-layer.
Alternatively `bg-layer` could work, `bg` being abbreviation for background.

**Comment by aagonzales** [comment]
# Alpha release final names
- [View full list of tokens](https://www.notion.so/carbondesignsystem/Final-List-of-new-tokens-fc5985c7331341baa6546064399d36da)
- **Conclusion**: we decided on a bit of a hybrid of naming conventions based the feedback we were receiving (thanks everyone for that feedback btw). We are keeping the names as simple and straight forward as possible, not adding extra naming just for the sake of a convention. 
- **Convention**: [element]-[role]-[style modifier]-[state]
![image](https://user-images.githubusercontent.com/11670886/113314924-629a5100-92d2-11eb-9d87-5b1a1fa892d1.png)

**Comment by aagonzales** [comment]
**Token names I'm still unsure about:**
- layer-selected-inverse
- border-inverse
- layer-alt

**Reply by tw15egan** [reply]
`layer-alt` was also one that stood out to me when working on the token updates. Not sure on a good alternative,`layer-contrast`...?

**Reply by aagonzales** [reply]
`layer-accent` maybe? Yeah its not really an "alternate" to layer. It's more of a supporting color.

**Reply by tw15egan** [reply]
I like `layer-accent`

**Reply by laurenmrice** [reply]
I like `layer-accent`!

**Reply by tw15egan** [reply]
Do we want to get any more feedback on this one? Or should I go ahead and update the token in the theme package?

**Reply by aagonzales** [reply]
Let's go ahead and update it!
